sdk/go/common/resource/testing/rapid.go

```diff
@@ -0,0 +1,226 @@
+func BoolPropertyGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.PropertyValue {
+		return resource.NewBoolProperty(rapid.Bool().Draw(t, "booleans").(bool))
+	})
+}
+
+// NumberPropertyGenerator generates numeric resource.PropertyValues.
+func NumberPropertyGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.PropertyValue {
+		return resource.NewNumberProperty(rapid.Float64().Draw(t, "numbers").(float64))
+	})
+}
+
+// StringPropertyGenerator generates string resource.PropertyValues.
+func StringPropertyGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.PropertyValue {
+		return resource.NewStringProperty(rapid.String().Draw(t, "strings").(string))
+	})
+}
+
+// TextAssetGenerator generates textual *resource.Asset values.
+func TextAssetGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) *resource.Asset {
+		asset, err := resource.NewTextAsset(rapid.String().Draw(t, "text asset contents").(string))
+		require.NoError(t, err)
+		return asset
+	})
+}
+
+// AssetGenerator generates *resource.Asset values.
+func AssetGenerator() *rapid.Generator {
+	return TextAssetGenerator()
+}
+
+// AssetPropertyGenerator generates asset resource.PropertyValues.
+func AssetPropertyGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.PropertyValue {
+		return resource.NewAssetProperty(AssetGenerator().Draw(t, "assets").(*resource.Asset))
+	})
+}
+
+// LiteralArchiveGenerator generates *resource.Archive values with literal archive contents.
+func LiteralArchiveGenerator(maxDepth int) *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) *resource.Archive {
+		var contentsGenerator *rapid.Generator
+		if maxDepth > 0 {
+			contentsGenerator = rapid.MapOfN(rapid.StringMatching(`^(/[^[:cntrl:]/]+)*/?[^[:cntrl:]/]+$`), rapid.OneOf(AssetGenerator(), ArchiveGenerator(maxDepth-1)), 0, 16)
+		} else {
+			contentsGenerator = rapid.Just(map[string]interface{}{})
+		}
+		archive, err := resource.NewAssetArchive(contentsGenerator.Draw(t, "literal archive contents").(map[string]interface{}))
+		require.NoError(t, err)
+		return archive
+	})
+}
+
+// ArchiveGenerator generates *resource.Archive values.
+func ArchiveGenerator(maxDepth int) *rapid.Generator {
+	return LiteralArchiveGenerator(maxDepth)
+}
+
+// ArchivePropertyGenerator generates archive resource.PropertyValues.
+func ArchivePropertyGenerator(maxDepth int) *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.PropertyValue {
+		return resource.NewArchiveProperty(ArchiveGenerator(maxDepth).Draw(t, "archives").(*resource.Archive))
+	})
+}
+
+// ResourceReferenceGenerator generates resource.ResourceReference values.
+func ResourceReferenceGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.ResourceReference {
+		return resource.ResourceReference{
+			URN:            URNGenerator().Draw(t, "referenced URN").(resource.URN),
```

Comment:
I'd worry about implicit empty values like "" if they're allowed/can occur in any of these fields, we'd need to prod rapid to make them appear more often. Random strings shrink to "" but it might miss them easily, and regex-driven strings don't cover "" if it does not match.

Reply:
> I'd worry about implicit empty values like "" if they're allowed/can occur in any of these fields, we'd need to prod rapid to make them appear more often. Random strings shrink to "" but it might miss them easily, and regex-driven strings don't cover "" if it does not match.

Absolutely. I think that for the most part we're in the clear with these generators. The one exception that comes to mind is the `parentType` portion of a URN, which is allowed to be empty but will currently always be a legal type token. I'll adjust that.